File: packages/expo-targets/src/modules/storage/ios/ExpoTargetsStorageModule.swift
import ExpoModulesCore
import WidgetKit

public class ExpoTargetsStorageModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoTargetsStorage")

    Function("setInt") { (key: String, value: Int, suite: String?) -> Void in
      let defaults = UserDefaults(suiteName: suite ?? "")
      defaults?.set(value, forKey: key)
      defaults?.synchronize()
    }

    Function("setString") { (key: String, value: String, suite: String?) -> Void in
      let defaults = UserDefaults(suiteName: suite ?? "")
      defaults?.set(value, forKey: key)
      defaults?.synchronize()
    }

    Function("setObject") { (key: String, value: [String: Any], suite: String?) -> Bool in
      let defaults = UserDefaults(suiteName: suite ?? "")
      defaults?.set(value, forKey: key)
      defaults?.synchronize()
      return true
    }

    Function("get") { (key: String, suite: String?) -> String? in
      let defaults = UserDefaults(suiteName: suite ?? "")
      return defaults?.string(forKey: key)
    }

    Function("remove") { (key: String, suite: String?) -> Void in
      let defaults = UserDefaults(suiteName: suite ?? "")
      defaults?.removeObject(forKey: key)
      defaults?.synchronize()
    }

    Function("refreshTarget") { (name: String?) -> Void in
      // Refresh widgets
      if #available(iOS 14.0, *) {
        if let targetName = name {
          WidgetCenter.shared.reloadTimelines(ofKind: targetName)
        } else {
          WidgetCenter.shared.reloadAllTimelines()
        }
      }

      // Refresh controls (iOS 18+)
      if #available(iOS 18.0, *) {
        if let targetName = name {
          ControlCenter.shared.reloadControls(ofKind: targetName)
        } else {
          ControlCenter.shared.reloadAllControls()
        }
      }
    }
  }
}

File: packages/expo-targets/android/src/main/java/expo/modules/targets/ExpoTargetsStorageModule.kt
package expo.modules.targets

import android.content.Context
import android.content.SharedPreferences
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONObject

class ExpoTargetsStorageModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoTargetsStorage")
        
        // iOS-compatible function signature: key, value, suite (widgetName)
        AsyncFunction("setInt") { key: String, value: Int, widgetName: String? ->
            getPreferences(widgetName ?: "default").edit().putInt(key, value).apply()
        }
        
        AsyncFunction("setString") { key: String, value: String, widgetName: String? ->
            getPreferences(widgetName ?: "default").edit().putString(key, value).apply()
        }
        
        AsyncFunction("setObject") { key: String, value: Map<String, Any>, widgetName: String? ->
            val jsonObject = JSONObject()
            value.forEach { (k, v) ->
                when (v) {
                    is String -> jsonObject.put(k, v)
                    is Number -> jsonObject.put(k, v)
                    is Boolean -> jsonObject.put(k, v)
                    else -> jsonObject.put(k, v.toString())
                }
            }
            val jsonString = jsonObject.toString()
            getPreferences(widgetName ?: "default").edit().putString(key, jsonString).apply()
            return@AsyncFunction true
        }
        
        AsyncFunction("get") { key: String, widgetName: String? ->
            return@AsyncFunction getPreferences(widgetName ?: "default").getString(key, null)
        }
        
        AsyncFunction("remove") { key: String, widgetName: String? ->
            getPreferences(widgetName ?: "default").edit().remove(key).apply()
        }
        
        AsyncFunction("getAllKeys") { widgetName: String? ->
            return@AsyncFunction getPreferences(widgetName ?: "default").all.keys.toList()
        }
        
        AsyncFunction("getAllData") { widgetName: String? ->
            val prefs = getPreferences(widgetName ?: "default")
            val result = mutableMapOf<String, Any>()
            prefs.all.forEach { (key, value) ->
                result[key] = value ?: ""
            }
            return@AsyncFunction result
        }
        
        AsyncFunction("clearAll") { widgetName: String? ->
            getPreferences(widgetName ?: "default").edit().clear().apply()
        }
        
        AsyncFunction("refreshTarget") { targetName: String? ->
            if (targetName != null) {
                ExpoTargetsReceiver.refreshWidget(appContext.reactContext!!, targetName)
            }
        }
        
        AsyncFunction("getTargetsConfig") {
            // Android doesn't have Info.plist equivalent, return null
            return@AsyncFunction null
        }
    }
    
    private fun getPreferences(widgetName: String): SharedPreferences {
        return appContext.reactContext?.getSharedPreferences(
            "expo_targets_$widgetName",
            Context.MODE_PRIVATE
        ) ?: throw Exception("Context not available")
    }
}
